tests/e2e/tidbcluster/serial.go

```diff
@@ -1008,4 +1008,131 @@ var _ = ginkgo.Describe("[tidb-operator][Serial]", func() {
+				}
+				podList, err := c.CoreV1().Pods(ns).List(listOptions)
+				if err != nil {
+					return nil, err
+				}
+				return podList.Items, nil
+			}
+
+			tc, err := cli.PingcapV1alpha1().TidbClusters(ns).Get(tcName, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to get tc")
+
+			pdPods, err := getPods(labels.SelectorFromSet(label.New().Instance(tcName).PD().Labels()).String())
+			framework.ExpectNoError(err, "failed to get pd pods")
+
+			tikvPods, err := getPods(labels.SelectorFromSet(label.New().Instance(tcName).TiKV().Labels()).String())
+			framework.ExpectNoError(err, "failed to get tikv pods")
+
+			tidbPods, err := getPods(labels.SelectorFromSet(label.New().Instance(tcName).TiDB().Labels()).String())
+			framework.ExpectNoError(err, "failed to get tidb pods")
+
+			// Upgrade CRD / Operator to current version
+			ocfg.Tag = cfg.OperatorTag
+			ocfg.Image = cfg.OperatorImage
+			oa.InstallCRDOrDie(ocfg)
+			oa.UpgradeOperatorOrDie(ocfg)
+			err = wait.Poll(5*time.Second, 10*time.Minute, func() (done bool, err error) {
+
+				newTc, err := cli.PingcapV1alpha1().TidbClusters(ns).Get(tcName, metav1.GetOptions{})
+				if err != nil {
+					return false, nil
+				}
+				// wait tidb to be updated
+				if tc.Status.TiDB.StatefulSet.CurrentRevision == newTc.Status.TiDB.StatefulSet.CurrentRevision {
+					klog.Info("wait tidb to be updated")
+					return false, nil
+				}
+				// wait tidb finish updating
+				if newTc.Status.TiDB.StatefulSet.CurrentRevision != newTc.Status.TiDB.StatefulSet.UpdateRevision {
+					klog.Info("wait tidb finish updating")
+					return false, nil
+				}
+
+				// confirm the tidb pod have been changed
+				changed, err := utilpod.PodsAreChanged(c, tidbPods)()
+				if changed {
+					klog.Infof("confirm tidb pods have been changed")
+				} else {
+					if err != nil {
+						klog.Errorf("meet error during verify tidb pods, err:%v", err)
+						return false, nil
+					}
+					if !changed {
+						return false, fmt.Errorf("tidb should be updated after operator upgrading")
+					}
+				}
+
+				// confirm the pd Pod haven't been changed
+				changed, err = utilpod.PodsAreChanged(c, pdPods)()
+				if err != nil {
+					klog.Errorf("meet error during verify pd pods, err:%v", err)
+					return false, nil
+				}
+				if changed {
+					return false, fmt.Errorf("pd pods have been changed after upgrading operator")
+				}
+				klog.Infof("confirm pd pods haven't been changed")
+
+				// confirm the pd tikv haven't been changed
+				changed, err = utilpod.PodsAreChanged(c, tikvPods)()
+				if err != nil {
+					klog.Errorf("meet error during verify tikv pods, err:%v", err)
+					return false, nil
+				}
+				if changed {
+					return false, fmt.Errorf("tikv pods have been changed after upgrading operator")
+				}
+				klog.Infof("confirm tikv pods haven't been changed")
+
+				return true, nil
+			})
+			framework.ExpectNoError(err, "Failed to check TidbCluster Status After Upgrading Operator")
```

Comment:
we should wait for pd/tikv pods are not changed in a certain of time, the expected err should be `wait.ErrWaitTimeout`, example:
https://github.com/pingcap/tidb-operator/blob/a19b72eb7cbf79f3ce77ff00d4a74547e05615d2/tests/e2e/tidbcluster/stability.go#L162-L189

`10*time.Minute` can be changed to `5*time.Minute`.  5 minutes is enough to make sure our logic is correct if pd/tikv pods are not affected in this time.

before checking pd/tikv pods are not affected, you can wait for tidb pods are affected. in this checking, what you expect is `nil` err (return `true, nil` when the tidb pods are different).

Reply:
I think the current logic is like what you said. We would expect tidb pods updated yet, then we expect the tikv and pd pods haven't been changed. The whole process may need 5 or 10 minutes.

Reply:
you didn't check pd/tikv pods are not affected in a certain time.  after tidb pods are changed, the wait function checks pd/pods once and return `true, nil` if they are not changed. what we expect is they are not changed in a certain time (e.g. 5 minutes).

Reply:
you can have two `wait.Poll` functions:

1) the first waits for the tidb pods are changed
2) the second waits for the pd/tikv pods are not changed in 5 minutes

Reply:
good idea, updated.